src/httpx/asyncstream.nim

```diff
@@ -0,0 +1,392 @@
+): AsyncStream[T] =
+  ## Creates a new AsyncStream.
+  ## 
+  ## Arguments:
+  ##  - `maxQueueLen`: The maximum queue length to use
+  ##  - `afterReadCb`: Optional callback to be called after every read (defaults to None)
+  ##  - `afterWriteCb`: Optional callback to be called after every write (defaults to None)
+
+  result = new AsyncStream[T]
+  result.queue = initDeque[T](maxQueueLen)
+  result.maxQueueLenInternal = maxQueueLen
+  result.readCbs = initDeque[AsyncStreamCb](1)
+  result.writeCbs = initDeque[AsyncStreamCb](1)
+  result.afterReadCb = afterReadCb
+  result.afterWriteCb = afterWriteCb
+  result.isCompletedInternal = false
+  result.exceptionInternal = none[ref Exception]()
+
+func isCompleted*[T](this: AsyncStream[T]): bool {.inline.} =
+  ## Whether the stream is completed.
+  ## If true, exception will be None.
+  
+  return this.isCompletedInternal
+
+func isFailed*[T](this: AsyncStream[T]): bool {.inline.} =
+  ## Whether the stream is failed.
+  ## If true, exception will be Some.
+  
+  return this.exceptionInternal.isSome
+
+func isFinished*[T](this: AsyncStream[T]): bool {.inline.} =
+  ## Whether the stream is either completed or failed.
+  ## To check each individually, use isCompleted and isFailed, respectively.
+  
+  return this.isCompleted or this.isFailed
+
+func exception*[T](this: AsyncStream[T]): Option[ref Exception] {.inline.} =
+  ## The exception that was the cause of the stream failing to be written/read, or none if there was no exception.
+  ## If an exception is present, any reads to the stream when the queue is empty will raise this exception.
+  ## The isFinished property should be true if this is some.
+  
+  return this.exceptionInternal
+
+func queueLen*[T](this: AsyncStream[T]): int {.inline.} =
+  ## The current internal queue length
+
+  return this.queue.len
+
+func maxQueueLen*[T](this: AsyncStream[T]): int {.inline.} =
+  ## The maximum queue length
+  
+  return this.maxQueueLenInternal
+
+func queue*[T](this: AsyncStream[T]): Deque[T] {.inline.} =
+  ## The internal queue.
+  ## 
+  ## Warning: modifying the queue manually while there are pending reads or writes will cause a deadlock until a new read/write occurs.
+  ## Exercise caution when modifying it!
+
+proc `maxQueueLen=`*[T](this: AsyncStream[T], newLen: range[1..high(int)]) {.inline.} =
+  ## Sets the maximum queue length.
+  ## Raises ValueError if the new length is less than the current queue length.
+  ## Lowering queue size while a stream is active is dangerous because it can lower stream performance.
+  ## Changing maximum queue length is expensive because it will result in memory allocation or deallocation.
+  
+  if unlikely(newLen < this.queue.len):
+    raise newException(ValueError, "Cannot lower max queue length lower than the current queue size")
+
+  this.maxQueueLenInternal = newLen
+
+proc write*[T](this: AsyncStream[T], item: sink T, prepend: bool|void = void): Future[void] =
```

Comment:
`void` can introduce lots of bugs. It is better to use default parameters or Options.

Reply:
I used void so that I could optimize with `when`, but using `Options` would probably be safer. I'll look into some more changes to it this week, I'm not very happy with how some parts are right now.

Reply:
also @ringabout , I've noticed that httpx performed slower than httpbeast before my changes, is this because of the selectors library? There aren't many differences between them. On my machine, httpbeast got 300k requests per second, httpx got about 100k requests per second. I'm wondering if that has something to do with HTTP pipelining (which is useless in the real world).

Reply:
> I'm wondering if that has something to do with HTTP pipelining (which is useless in the real world).

Probably, the differences I know of are that httpx removed HTTP pipeline and perhaps httpeast applies some optimization like https://github.com/dom96/httpbeast/commit/658685d7954d0e3cfe92550adf7ea644b1a508dc

Reply:
Ok, I'm glad you removed that. It adds unnecessary complexity. I wasn't sure though, since there were still comments in the code talking about pipelining. I need to check whether my load testing tool uses pipelining or not.